### ISSUE TITLE
Error adding new node

### ISSUE THREAD

**Issue by drmrbrewer** [githubsaturn/captainduckduck#122]
I'm trying to add a new node (only one at present), based on the guide [here](https://github.com/githubsaturn/captainduckduck/wiki/Server-Clustering-and-App-Scaling).

In Captain dashboard on the first node, I've clicked “Enable Self Hosted Registry” as instructed.

I've entered all the IP address details as instructed on the Nodes page too, stating "worker" for type and "root" for user.  I've added a private key from the second server too.  Then I click Join Cluster, and all I get is a red popup error in the top corner saying "Error, Please try again.."

Any ideas?  On the second server I installed Docker, but not Captain... is that right?

**Comment by githubsaturn** [comment]
Do you happen to have an enabled firewall? Simply run `ufw disable` on your main node and on your second note. If you're using ubuntu to disable the firewall. more details on firewall is on [wiki](https://github.com/githubsaturn/captainduckduck/wiki/Firewall)

If this doesn't solve the issue, ssh to your server and run
```
docker service logs captain-captain --since 10m --follow 
```

While you are viewing logs, try to reproduce the error and now attach the logs to this issue so I can take a closer look at it.

**Comment by drmrbrewer** [comment]
OK I got an error about an "encrypted key but no passphrase given".  So that reminded that when I created the ssh key on server 2, I used a passphrase.  So I generated another, without passphrase, and used the private key from that.  This time the error on creating cluster is about an SSH error.  The logs on the main server during that error contains:

	| January 13th 2018, 6:50:30.741 pm    Error: All configured authentication methods failed
	| Error: All configured authentication methods failed
	|     at tryNextAuth (/usr/src/app/node_modules/ssh2/lib/client.js:380:17)
	|     at SSH2Stream.onUSERAUTH_FAILURE (/usr/src/app/node_modules/ssh2/lib/client.js:591:5)
	|     at emitTwo (events.js:125:13)
	|     at SSH2Stream.emit (events.js:213:7)
	|     at parsePacket (/usr/src/app/node_modules/ssh2-streams/lib/ssh.js:3929:10)
	|     at SSH2Stream._transform (/usr/src/app/node_modules/ssh2-streams/lib/ssh.js:669:13)
	|     at SSH2Stream.Transform._read (_stream_transform.js:186:10)
	|     at SSH2Stream._read (/usr/src/app/node_modules/ssh2-streams/lib/ssh.js:251:15)
	|     at SSH2Stream.Transform._write (_stream_transform.js:174:12)
	|     at doWrite (_stream_writable.js:387:12)
	| January 13th 2018, 6:50:30.746 pm    Error: SSH Connection error!!
	| Error: SSH Connection error!!
	|     at Object.createError (/usr/src/app/src/api/ApiStatusCodes.js:5:21)
	|     at Client.<anonymous> (/usr/src/app/src/user/CaptainManager.js:635:51)
	|     at emitOne (events.js:115:13)
	|     at Client.emit (events.js:210:7)
	|     at tryNextAuth (/usr/src/app/node_modules/ssh2/lib/client.js:382:12)
	|     at SSH2Stream.onUSERAUTH_FAILURE (/usr/src/app/node_modules/ssh2/lib/client.js:591:5)
	|     at emitTwo (events.js:125:13)
	|     at SSH2Stream.emit (events.js:213:7)
	|     at parsePacket (/usr/src/app/node_modules/ssh2-streams/lib/ssh.js:3929:10)
	|     at SSH2Stream._transform (/usr/src/app/node_modules/ssh2-streams/lib/ssh.js:669:13)
	| POST /api/v1/user/system/nodes/ 200 72.646 ms - 54

**Comment by drmrbrewer** [comment]
Once the second node is added, do you always have to redeploy your apps (I haven't done that since enabling the registry)?

What is the easiest way to confirm that the second node is active and handling requests?  Does it show in the netdata monitor?

**Comment by githubsaturn** [comment]
Glad that you figured it out.

> What is the easiest way to confirm that the second node is active and handling requests? Does it show in the netdata monitor?

NetData only works for the leader node. It's a Docker shortcoming: https://github.com/moby/moby/issues/25885

The easiest way to test is to create a test app and deploy the sample NodeJS app, that app shows the public IP address of the server on which it is sitting. After deploying the sample app, you can increase the instance count to 3 or 4 and keep refreshing the page. You will see different IP addresses appearing. 


> Once the second node is added, do you always have to redeploy your apps (I haven't done that since enabling the registry)?

Yes - it's needed because the current image only sit on the leader node, apps cannot start on the second node. Once you reploy (after the registry is enabled). Images will be pushed to registry and therefore will be available to other nodes as well.

**Comment by drmrbrewer** [comment]
> NetData only works for the leader node

So how can I monitor performance of the second node?  Can I install Captain on the second node (presently I have not done this... didn't think it was necessary when just adding it as a worker to the first node)... and then access netdata for the second node via the Captain control panel interface?

**Comment by githubsaturn** [comment]
> Can I install Captain on the second node

No, having two managers will mess up with container management. 

Leader node is what you most care about. But for sure, there are ways you can install netdata on the second node as well. The difficulty comes in if you want to make it as secure as the netdata on the leader node, you have to do all the things that Captain does for you manually: starting an nginx container, starting a let's encrypt container, starting the netdata container. It's just extremely difficult. 

To install unsecured netdata temporarily and remove it (for dev purpose), on your second node, run
```
docker run -d --cap-add SYS_PTRACE \
           --rm \
           --name netdata-container \
           -v /proc:/host/proc:ro \
           -v /sys:/host/sys:ro \
           -p 19999:19999 titpetric/netdata
```
then access netdata on your second node via `http://IPADDRESS:19999`. Once you are done, run:
```
docker rm netdata-container --force
```

PS: Don't forget to thumbs up the Docker issue that I linked. If they fixed that limitation, we can install netdata through Captain on other nodes.

**Comment by drmrbrewer** [comment]
> thumbs up the Docker issue that I linked

done!

> If they fixed that limitation, we can install netdata through Captain on other nodes

That would indeed be the best solution.  Thanks for your time (again) in explaining all of these things so clearly.

**Comment by drmrbrewer** [comment]
P.S. regarding confirming that the apps on the second node are active and handling traffic, I note that if I do the following on the first node:

    $ docker service logs srv-captain--myapp --since 10m --follow

the logs seems to show entries from all apps (on both nodes)?  I see output like the following... I have two nodes and four instances, and I see `myapp.1` to `myapp.4`, and also `server-01` and `server-02`:

	srv-captain--myapp.3.l3n6ar2ykx3f@server-02 ---log output---
	srv-captain--myapp.1.l8m9cn4qkthr@server-01 ---log output---

**Comment by githubsaturn** [comment]
Yes that's correct. Logs are aggregated from all nodes.

**Comment by drmrbrewer** [comment]
Couple of observations from checking out what's happening on the second node.

First, it doesn't seem possible to view logs of instances on the second nodes, as it is on the first:

    $ docker service logs srv-captain--myapp --since 10m --follow
    Error response from daemon: This node is not a swarm manager. Worker nodes can't be used to view or modify cluster state. Please run this command on a manager node or promote the current node to a manager.

Secondly, it seems that the "steal" part of the CPU usage is quite high on the second node... certainly much higher than on the first node:

![cpu-steal](https://user-images.githubusercontent.com/9009792/35187539-562ef170-fe1d-11e7-868f-000a88dd245d.jpg)

Is this expected / normal?

**Comment by githubsaturn** [comment]
doesn't seem possible to view logs of instances

Correct. Logs from services are viewable from the manager node. Read more on services and containers if you want to know how they are related..

>  however I need to check whether this is at all expected for a worker node in a docker swarm pair.

I cannot comment on this. Personally, I think it's all in your VPS providers hand. But feel free to post this question on Docker forum / issue.

**Comment by drmrbrewer** [comment]
ah ok so logs are aggregated across all instances, even those on the other nodes.

as for the "steal" values, I've asked the provider to comment.

**Comment by drmrbrewer** [comment]
Definitely seems to be an issue with the second server... I added a third and that shows low steal values.  Interesting.

**Comment by mukeshwani** [comment]
Hello. I'm trying to add a second node. I might be missing something or could be a bug...

Here are my steps:
1. created a second server, installed Docker (did not install CapRover)
2. Setup SSH keys etc... verified I can SSH from the First Server to the Second Server.
3. Disabled firewall on both servers.
4. Went to the Cluster page on First Server and Enabled Self-Hosted Registry 
5. Added Node - entered the information as required.
6. Got the message that NOde was added successfully

But I don't see the second node on the Cluster page. I refreshed and logged out and logged in, still don't see the second Node.
I SSH into the second node and try the docker swarm leave and it says that it is not part of a docker swarm.

Any help?
